actionpack/lib/action_controller/metal/request_forgery_protection.rb

```diff
@@ -240,6 +288,57 @@ def handle_unverified_request
       end
     end
 
+    class SessionStore
+      def fetch(request)
+        request.session[:_csrf_token]
+      end
+
+      def store(request, csrf_token)
+        request.session[:_csrf_token] = csrf_token
+      end
+
+      def reset(request)
+        request.session.delete(:_csrf_token)
+      end
+    end
+
+    class CookieStore
+      def initialize(cookie = :csrf_token)
+        @cookie_name = cookie
+      end
+
+      def fetch(request)
+        contents = request.cookie_jar.encrypted[@cookie_name]
+        return nil if contents.nil?
+
+        value = JSON.parse(contents)
+        return nil unless request.session.initially_empty? || value["session_id"]&.fetch("public_id") == request.session.id.public_id
```

Comment:
With the additional check, the following is now true:

 1. If the session was empty when we started processing the request, then use the CSRF token in the cookie, regardless of the session ID
 1. If the session was not empty, only use the CSRF token if the session ID in the cookie matches the session ID.

The first point covers the following cases:
 * When the CSRF token was generated, the request completed with an empty session, it is still empty on the subsequent request
 * When the CSRF token was generated, the request completed with a non-empty session, but that session was somehow cleared (i.e. cache flush) prior to the subsequent request.

The second point covers the following cases:
 * When the CSRF token was generated, the request completed with a non-empty session, and it is still non-empty on subsequent requests.
 * When the CSRF token was generated, the request completed with an empty session, but it is non-empty on a subsequent request -> This will be blocked, but I think that's ok since it doesn't make sense that something can be added to the session in-between requests.

Empty sessions are still open to attack in the following scenario:
 1. Attacker obtains a valid cookie
 2. Attacker waits for victim to log-in, and the user is in a state with an empty session
 3. Attacker overwrites the cookie with the prepared one
 4. Attacker uses a form-specific token obtained by requesting a form themselves

We're better off than we were before because this only works when both sessions are empty.  But I feel like we're back to question of whether it makes sense to protect empty sessions?  As we said earlier, the session is empty, so we're not really protecting it.  The only thing that we are protecting is data stored in cookies.

If it's a requirement to protect the empty session from the attack described here, then I'm not sure that any of this is possible.

Reply:
> Empty sessions are still open to attack in the following scenario:

> Attacker overwrites the cookie with the prepared one

I think that if the attacker can do that, then it can already bypass the existing session stored token. No? So I think it's fine.

>  whether it makes sense to protect empty sessions?

Yes it does. Otherwise you can use CSRF to force log someone into an account you control.